crates/apollo-router-core/src/traits.rs

```diff
@@ -55,6 +55,19 @@ where
 
 impl<T: ?Sized> WithCaching for T where T: QueryPlanner + Sized {}
 
+/// An object that accepts a [`Request`] and allow creating [`Route`]'s.
+pub trait Router<T: Route>: Send + Sync + Debug {
```

Comment:
I assume this is eventually going to allow us to support caching the preparation phase of a query. 

Possible renames:
`create_route` -> `prepare`
`Route` -> `PreparedQuery`

Also I would expect `execute` to take `request` as a parameter and not store it in `ApolloRoute`.

Reply:
> I assume this is eventually going to allow us to support caching the preparation phase of a query.

:100: 

> Also I would expect execute to take request as a parameter and not store it in ApolloRoute.

I think you're :100: right. The object itself could be cached and the variables of the request may vary.

> Possible renames:
> create_route -> prepare
> Route -> PreparedQuery

I was so proud to find the name `Router -> create_route() -> Route -> execute()` :disappointed:  But I realized during the review that all of you got very confused with that and as I'm not really a graphql user myself I will refer to your judgment of you all to see what is the better name for these. cc @o0Ignition0o @Geal

Reply:
😅 I think these bits will come together once query_plan and query_plan.execute() do!

Reply:
Fixed in 8f3eb4389cbb5ee08e11913074d84319e8354e2f and 03042ee37cad233345b674f97089fc1638111036